### ISSUE TITLE
Unique optimization values per pair

### ISSUE THREAD

**Issue by mosquitotrip** [freqtrade/freqtrade#6871]
Operating system: Debian
Python Version: 3.9.12
CCXT version: 1.83.10
Freqtrade Version: develop-d5486f17d

As every market has unique arithmetic and behavior it's illogical to assume that the same indicator settings will apply perfectly for a group of pairs.

Some crypto markets have a strong correlation, yet the actual performance of the indicators needs to be perfectly fitted (as possible) per market.

Therefore I would like to suggest the following:

1. As unique parameters were introduced with buy/sell hyperopt settings, introducing a per-pair flag (rather in boot or in config) to define rather FT optimizes a per-pair scenario
2. A per-pair scenario should be guided using {self.quote} or similar in strategy for proper identification of each dataframe
3. the saved hyperopt JSON file should include per-pair parameters
4. Backtesting should be transparent 

Having this feature I believe will greatly improve FT performance in live/dry trading

**Comment by xmatthias** [comment]
Optimizing indicators per pair will give you a perfect fit (assuming enough epochs).

While that sounds good in theory (all trades are now perfect) - that strategy would not perform well on unseen data - therefore i think this whole approach is not a good idea / approach - and will most likely fail the moment you attempt to run it on unseen data.


For a "per pair" optimized parameter - you'd also always have to use a static pairlist (otherwise you'd run with parameters from a random pair).
As then all pairs are known, you can pretty easily use "per pair" parameters directly in the strategy (just prefix/postfix the parameter name by pair). 
I don't see why this should not be accomplishable from within the strategy without changes to freqtrade - if you think this is something that'll give you more profits (i think it won't - and will fail on future data).


At the moment, I therefore don't see freqtrade complicating it's code in this area (reducing maintainability as a consequence) - as i 
a) don't see the benefit of this approach
b) don't see why it can't be accomplished in the strategy

if you think this is a must and there's things preventing the implementation in a strategy - you're obviously free to submit a Pull request adding this, assuming the code is well made, tested - and will not break existing functionality.

**Comment by mosquitotrip** [comment]
@xmatthias I've been running a hacked strategy with unique optimization settings for each pair against aggressive leveraged markets (20-30 pairs per bot instance), with quite simple strategy.

Here's some results:
**FTX:**
**Performance:**
1.  VETBULL/USD 7.485 USD (16.33%) (13)
2.  TRXBULL/USD 4.319 USD (9.77%) (2)
3.  ALTBULL/USD 3.25 USD (6.59%) (14)
4.  ADABULL/USD 3.115 USD (6.71%) (14)
5.  DRGNBULL/USD 2.825 USD (5.25%) (5)
6.  PRIVBULL/USD 2.603 USD (5.28%) (2)
7.  KNCBULL/USD 2.211 USD (4.56%) (2)
8.  ETCBULL/USD 1.707 USD (3.48%) (2)
9.  ATOMBULL/USD 0.952 USD (2.30%) (6)

**KuCoin:**
**Performance:**
1.  VET3L/USDT 29.925 USDT (25.80%) (11)
2.  UNI3L/USDT 26.068 USDT (34.16%) (8)
3.  XRP3L/USDT 24.011 USDT (13.68%) (5)
4.  AXS3L/USDT 15.949 USDT (9.74%) (5)
5.  AVAX3L/USDT 14.148 USDT (18.32%) (6)

Of course, not all pairs will perform as expected but that's more signals related, the idea is to find the most profitable pairs and NOT from backtesting results. We can only achieve this with unique parameters and inputs from actual real-time signals, it's logical.

The bottom line is obviously much better results with unique parameters per pair.

I wish I would have the coding skills to perform this, I can just suggest at this point the benefits as I see it performs better.

Do you think it can be accomplished through strategy without any code modifications? 

Maybe introducing hyperopt settings/parameters in lower levels within the strategy class can 'ease' on users to correlate hyperopt-settings and a quote, but that's just a thought ;)

**Comment by xmatthias** [comment]
Using leveraged pairs is usually misleading - as they almost always have very low volume - and performance will be market depending (all your FTX results are on BULL - but once market changes just a little bit, you'll have high losses on this). 
You'll have to have a strategy running live/dry for months to properly confirm such a thing - especially in these markets, as the shown profits above can be wiped in half a day - once market behaviour (speak trend) changes slightly.

Optimizing per pair will base your strategy on one (false) assumption - which is that markets continue to move in line with how they did move in the past.

> Do you think it can be accomplished through strategy without any code modifications?

Nothing prevents you from having a parameter `buy_VETBULL = ` ... - and then use the that when VETBULL/USD is the pair (you can get that via metadata).
it may be a bit of work (not really - you can quite easily define this for a given list of pairs) - and then only requires some magic to get the proper value


```python

class whatever(IStrategy):

    buy_rsi = IntParameter(10, 40, default=30, space="buy")

    def populate_entry_trend(self, dataframe, metadata):
       
       asset=metadata['pair'].split('/')[0] # not 100% reliable
      # buy_rsi becomes the fallback. your real parameters will be defined below
       val = getattr(self, f"buy_rsi_{asset}", self.buy_rsi)
       # signal assignment

# Careful, this must be AFTER the strategy - at the same level as "class" 
pairs = ['VETBULL', 'TRXBULL', ] # ...
for p in pairs:
    setattr(DynamicParams, f'buy_rsi_{p}', IntParameter(10, 40, default=30, space="buy"))

```

With the above example, i get 3 buy parameters (`buy_rsi`, `buy_rsi_VETBULL` and `buy_rsi_TRXBULL`).
We'll obviously want a "generic fallback" so it's not failing (you could handle this differently, like skipping the signal assignment if no dedicated parameter is found.

You'll obviously have to keep the pairlist here - but i don't see this as a problem, really, as you explicitly said you'd want "per pair" parameters, so using a dynamic pairlist (one you didn't hyperopt on) is out of the question anyway.

**Comment by mosquitotrip** [comment]
Very interesting, it is good to discuss practice, even without code modification.

So potentially a static pairlist can be achieved on the class level overriding a dynamic pairlist with unique parameters.

That's good enough ;)

**Comment by xmatthias** [comment]
your approach implicitly prohibits dynamic pairlists (you'd be missing parameters for the "new" pairs) :)

**Comment by mosquitotrip** [comment]
> your approach implicitly prohibits dynamic pairlists (you'd be missing parameters for the "new" pairs) :)

But potentially simply passing pairlist array in the class level can help keep dynamic pairlists and unique parameters/dataframes

With my suggested approach, every pair will have unique parameters saved per hyperopt session, your suggesting to unique the actual variables, that almost the same thing ;)

Accessing pairlist array on class level is something to consider?

**Comment by xmatthias** [comment]
> Accessing pairlist array on class level is something to consider?

No - this impossible.

**Comment by sanoj2021** [comment]
Hi, the possibility to test a strategy optimized for single pairs sounds promising for me.
I read about the method using multiple config files in #4736, but ...
Is this also possible by simply putting `if metadata['pair'] == 'X/Y'` for every pair in populate_indicators(), populate_entry_trend() and populate_exit_trend() like this?
```
def populate_indicators(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        if metadata['pair'] == 'BTC/USD':
            buy_BTC_rsi = ...
        if metadata['pair'] == 'ETH/USD':
            buy_ETH_rsi = ...
def populate_entry_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        if metadata['pair'] ==  'BTC/USD':
            buy_BTC_rsi logic here
        if metadata['pair'] ==  'ETH/USD':
            buy_ETH_rsi logic here
def populate_exit_trend analog...
```

**Comment by xmatthias** [comment]
it's a possibility - but will most likely result in a curve-fitted (or over-optimized) strategy against past performances - which usually performs quite poorly on unseen data.

**Comment by sanoj2021** [comment]
Thanks for the confirmation of the functionality of my code example.
I would really like to see the over-fitting myself and run tests on this respectively.

**Comment by sanoj2021** [comment]
I tried using the line
`setattr(DynamicParams, f'buy_rsi_{p}', IntParameter(10, 40, default=30, space="buy"))`
but `DynamicParams` is not defined and could not find it anywhere in freqtrade.
Would really like to use this and very much appreciate if - someone can show how and where to define `DynamicParams`?
